rust/src/sdk/api.rs

```diff
@@ -87,85 +84,174 @@ macro_rules! PrimModel {
-
-    #[repr(C)]
-    struct Input {
-        _round_started: bool,
-        _participant_initialized: bool,
-        _model_cached: bool,
+#[allow(unused_unsafe)]
+#[no_mangle]
+/// Starts the [`Client`].
+///
+/// The [`Client`]'s tasks are executed in an asynchronous [`Runtime`].
+///
+/// Takes a `callback(state, code)` function pointer and a void pointer to the `state` of the
+/// callback. The callback will be triggered when the client must be stopped because of a panic or
+/// error. See the [errors] section for the error `code` definitions.
+///
+/// # Errors
+/// Ignores null pointer `client`s and triggers the callback immediately. Triggers the callback
+/// with one of the following error codes in case of a [`Client`] panic or error:
+/// - `-1`: client didn't start due to null pointer
+/// - `0`: no error
+/// - `1`: client panicked due to unexpected/unhandled error
+/// - `2`: client stopped due to error [`ParticipantInitErr`]
+/// - `3`: client stopped due to error [`ParticipantErr`]
+/// - `4`: client stopped due to error [`DeserialiseErr`]
+/// - `5`: client stopped due to error [`GeneralErr`]
+///
+/// # Safety
+/// The method dereferences from the raw pointer arguments. Therefore, the behavior of
+/// the method is undefined if the arguments don't point to valid objects.
+///
+/// If the callback is triggered because of a panicking client, it is the users responsibility
+/// to not access possibly invalid state and to drop the client. If certain parts of the state
+/// can be guaranteed to be valid, they may be read before dropping.
+///
+/// [errors]: #errors
+/// [`ParticipantInitErr`]: ../../client/enum.ClientError.html#variant.ParticipantInitErr
+/// [`ParticipantErr`]: ../../client/enum.ClientError.html#variant.ParticipantErr
+/// [`DeserialiseErr`]: ../../client/enum.ClientError.html#variant.DeserialiseErr
+/// [`GeneralErr`]: ../../client/enum.ClientError.html#variant.GeneralErr
+pub unsafe extern "C" fn run_client(
+    client: *mut FFIClient,
+    callback: unsafe extern "C" fn(*mut c_void, c_int),
+    state: *mut c_void,
+) {
+    if client.is_null() {
+        return unsafe {
+            // safe if the `callback` is sound
+            callback(state, -1_i32 as c_int)
+        };
     }
-    let input = &Input {
-        _round_started: client.has_new_coord_pk_since_last_check,
-        _participant_initialized: true,
-        _model_cached: client.cached_model.is_some(),
-    } as *const _ as *const c_void;
+    let (runtime, client) = unsafe {
+        // safe if the raw pointer `client` comes from a valid allocation of a `FFIClient`
+        (&(*client).runtime, &mut (*client).client)
+    };
+    let code = match panic::catch_unwind(
+        // even though `&mut Client` is `!UnwindSafe` we can assert this because the user will be
+        // notified about a panic immediately to be able to safely act accordingly
+        panic::AssertUnwindSafe(|| runtime.handle().block_on(client.start())),
```

Comment:
I'm not sure about this. I don't fully grasp what `UnwindSafe` means and under which condition can we assert that `&mut Client` is `UnwindSafe`. But I'm a bit worried because:

- `Client.start` is very complex, so that makes the evaluation of unwind safety difficult
- even if we can prove that this is currently unwind safe, I'm worried we can make this unsafe without realizing it, by making changes to `Client` in the future.

So I have two questions:

- do we really need to `catch_unwind` here? Couldn't we just let it crash?
- should we add an `unsafe` block here?

Reply:
`UnwindSafe` basically says that there is no danger in accessing a value after a panic happened. this is generally true for immutable references, but not for mutable references.

currently i've put the note that it is the user's responsibility to not access possibly invalid values (we could clean up the `client` ourself but then the paradigm of create-use-destroy all happening on one side, in this case the non-Rust side of the API, is violated and might as well lead to severe bugs/segfaults).

the main issue is that we must catch the panic, because letting it propagate across the FFI-boundary is undefined behavior and will most likely result in segfaults. what we can do is to improve error handling on our side to reduce the number of possible panics and return proper errors instead. (for example the first panic that is catched here is a panic due to closed channels if the client is started without a coordinator service. things like that should of course be proper errors, but in the end we can't rule out every possible panic and need to have a way to catch them.)

`unsafe` is not needed for `AssertUnwindSafe` according to the docs, but i agree that we should enclose this in an `unsafe` block nonetheless. there is some hidden docs about unwind safety:
```
// Implementations of the `UnwindSafe` trait:
//
// * By default everything is unwind safe
// * pointers T contains mutability of some form are not unwind safe
// * Unique, an owning pointer, lifts an implementation
// * Types like Mutex/RwLock which are explicitly poisoned are unwind safe
// * Our custom AssertUnwindSafe wrapper is indeed unwind safe
```

Reply:
an interesting thing i just noticed is that `&mut T` is `!UnwindSafe` but `*mut T` is indeed `UnwindSafe`. i'm not sure why this is the case, but maybe we can leverage this since we are working with raw pointers most of the time anyways.

Reply:
I just asked about that, let's see: https://stackoverflow.com/questions/62422474/why-does-mut-t-implement-unwindsafe-but-mut-t-doesnt

Reply:
> UnwindSafe basically says that there is no danger in accessing a value after a panic happened. this is generally true for immutable references, but not for mutable references.

> currently i've put the note that it is the user's responsibility to not access possibly invalid values (we could clean up the client ourself but then the paradigm of create-use-destroy all happening on one side, in this case the non-Rust side of the API, is violated and might as well lead to severe bugs/segfaults).

> the main issue is that we must catch the panic, because letting it propagate across the FFI-boundary is undefined behavior and will most likely result in segfaults. what we can do is to improve error handling on our side to reduce the number of possible panics and return proper errors instead. (for example the first panic that is catched here is a panic due to closed channels if the client is started without a coordinator service. things like that should of course be proper errors, but in the end we can't rule out every possible panic and need to have a way to catch them.)

Thanks that clarifies everything. I think it would be nice to include these explanations in comments.

Reply:
i'll those as comments